scipy/ndimage/filters.py

```diff
@@ -139,19 +139,23 @@ def _gaussian_kernel1d(sigma, order, radius):
     """
     if order < 0:
         raise ValueError('order must be non-negative')
-    p = numpy.polynomial.Polynomial([0, 0, -0.5 / (sigma * sigma)])
-    x = numpy.arange(-radius, radius + 1)
-    phi_x = numpy.exp(p(x), dtype=numpy.double)
-    phi_x /= phi_x.sum()
-    if order > 0:
-        q = numpy.polynomial.Polynomial([1])
-        p_deriv = p.deriv()
-        for _ in range(order):
-            # f(x) = q(x) * phi(x) = q(x) * exp(p(x))
-            # f'(x) = (q'(x) + q(x) * p'(x)) * phi(x)
-            q = q.deriv() + q * p_deriv
-        phi_x *= q(x)
-    return phi_x
+    exponent_range = numpy.arange(order + 1)
+    sigma2 = sigma * sigma
+    q_coeffs = numpy.zeros(order + 1)
+    q_coeffs[0] = 1
+    for _ in range(order):
+        # f(x) = q(x) * phi(x) = q(x) * exp(p(x))
+        # f'(x) = (q'(x) + q(x) * p'(x)) * phi(x)
+        # p'(x) = -1 / sigma ** 2
+        new_coeffs = numpy.zeros(order + 1)
+        new_coeffs[:-1] = (q_coeffs * exponent_range)[1:]
+        new_coeffs[1:] += q_coeffs[:-1] / -sigma2
+        q_coeffs = new_coeffs
+    x = numpy.arange(-radius, radius+1)
+    phi_x = numpy.exp(-0.5 / sigma2 * x ** 2)
+    phi_x = phi_x / phi_x.sum()
+    q = (x[:, None] ** exponent_range) @ q_coeffs
```

Comment:
Feels awkward indeed but on paper, we dropped support and this should be OK. We removed the Py2 CI runs already.

Reply:
We dropped 2.7 and 3.4 in master already, and this is new code (enhancement, won't be backported) so `@` is okay.

Reply:
Oops, I forgot that @ isn't defined in Python < 3.5. It could be replaced with `dot` here but I'll leave it if it's not a concern :)

@ev-br, I'm not as familiar with the stability problems that arise from polynomials. Do you see any common gotchas in how this is implemented?

For readability, I wonder if it's better to define matrix operators for `q'(x)` and `q(x) * p(x)` instead of doing this with indexing. Something like:

    D = np.diag(np.arange(1, order), 1)
    P_prime = np.diag(np.ones(order) * -1 / sigma2, -1)
    for _ in range(order):
        q_coeffs = D @ q_coeffs + P_prime @ q_coeffs

I'll implement this and check speed, tests, etc. but IMO it's more clear what's happening inside the loop this way.

Reply:
We only support 3.5+ for new features so as @rgommers said `@` should be fine. FWIW it's already appearing in other PRs to SciPy (of mine, at least).

As far as the matrix operations, I'd say do whichever is fastest, and if they are the same, whichever is more readable. This is not so many lines of code so it shouldn't be too hard to understand. You can also write:
```
# The following is equivalent to:
# for ...
#    ...
... more complicated code ...
```
Then people can more easily understand what's going on but we get the speed benefits, too.

Reply:
Matrix operations ended up being faster, particularly for higher orders. Also, as it turns out `dot` is faster than `@`.

    Matrix
    ======
    %timeit _gaussian_kernel1d(5, 2, 20)
    27.4 µs ± 669 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

    %timeit _gaussian_kernel1d(5, 20, 20)
    44.8 µs ± 438 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

    Indexing
    ========
    %timeit _gaussian_kernel1d(5, 2, 20)
    28.6 µs ± 1.33 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

    %timeit _gaussian_kernel1d(5, 20, 20)
    135 µs ± 12 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Reply:
Ok, indeed. Sorry for the noise.

Reply:
Marking this comment as resolved since it was mostly about the @ operator.